kernel/gic/src/gic/mod.rs

```diff
@@ -50,24 +50,22 @@ bitflags! {
 
 /// Target of a shared-peripheral interrupt
 pub enum SpiDestination {
-    /// That interrupt must be handled by
-    /// a specific PE in the system.
+    /// The interrupt must be delivered to a specific CPU.
     Specific(CpuId),
-    /// That interrupt can be handled by
-    /// any PE that is not busy with another,
-    /// more important task
+    /// That interrupt can be handled by any PE that is not busy with another, more
+    /// important task
```

Comment:
hmm, this makes it sound like if one of the other PEs is busy with another more important task, the interrupt won't actually be handled/delivered at all. Is that correct, i.e., is it guaranteed to be delivered to just a single PE? Will that chosne PE eventually handle that interrupt once it has completed it's current "higher priority" task?

Let's not use the phrase "more important", that's quite vague; use "higher priority" or whatever the most correct and specific term is.

It'd also be nice to explain a bit about how the PE is actually chosen. Is it simply chosen by hardware? And is it guaranteed to be handled by just _one_ PE? I only ask because x86 has different behavior; typically you can have a single interrupt get delivered to multiple CPUs simultaneously, which makes it really hard to handle correctly.

Reply:
If I'm allowed to chime in:
This variant (AnyCpuAvailable, also named Mode 1) routes the interrupts to any connected PE that is participating in distribution of the interrupt's group. It's chosen by the Distributor (hardware) and can vary each time.
Haven't read about priority here, I do assume that's explained somewhere else.

GICv2 did send it to multiple targets, with the first to respond getting it (see below).

Reply:
From the GICv3 architecture specification:

![image](https://user-images.githubusercontent.com/61582713/228876617-10394a4c-da1f-4d81-a66e-e54f33bde48c.png)

- First condition = if the redistributor of that PE was initialized
- E1NWF stands for "Enable 1 of N Wakeup Functionality"; it's a bit in the control reg of the distributor (one per system)
- 3rd condition = whether the redistributor of that PE is configure to accept the interrupt group of that interrupt
- (4th condition, mentioned elsewhere = if the CPU interface is configured to accept the interrupt group of that interrupt)

In Theseus currently:
- 1st condition is asserted for the BSP only
- 2nd condition is not asserted (although we never put any core to sleep atm)
- 3rd condition is not asserted
- 4th condition: configured to accept group 1 interrupts only

I used to think that the redistributor is only relevant to LPIs; that was wrong.

Regarding which of the selectable PE is actually selected, it is implementation defined, so the running priority level is actually not necessarily involved in the choice:

![image](https://user-images.githubusercontent.com/61582713/228881610-b31717d7-c2d4-4142-a209-c98e4bdfa7d8.png)

I'll try to implement assertions for these conditions, even though we have no SPI to test this with currently.

Reply:
We can resolve that conversation now that these assertions were implemented by #915 :ok_hand: